src/fold.rs

```diff
@@ -418,21 +434,29 @@ fn cast(expr: Expr, ctype: &Type) -> CompileResult<ExprType> {
 /// and we can use that to store the new type
 fn const_cast(token: &Literal, ctype: &Type) -> Option<Literal> {
     let token = match (token, ctype) {
-        (Int(i), Type::Bool) => Int((*i != 0) as i64),
+        (Int(i), Type::Bool) => Int((*i != 0).into()),
         (Int(i), Type::Double) | (Int(i), Type::Float) => Float(*i as f64),
         (Int(i), ty) if ty.is_integral() && ty.is_signed() => Int(*i),
         (Int(i), ty) if ty.is_integral() => UnsignedInt(*i as u64),
-        (UnsignedInt(u), Type::Bool) => Int((*u != 0) as i64),
+
+        (UnsignedInt(u), Type::Bool) => Int((*u != 0).into()),
         (UnsignedInt(u), Type::Double) | (UnsignedInt(u), Type::Float) => Float(*u as f64),
         (UnsignedInt(u), ty) if ty.is_integral() && ty.is_signed() => Int(*u as i64),
         (UnsignedInt(u), ty) if ty.is_integral() => UnsignedInt(*u),
+
         (Float(f), Type::Bool) => Int((*f != 0.0) as i64),
         (Float(f), Type::Double) | (Float(f), Type::Float) => Float(*f),
         (Float(f), ty) if ty.is_integral() && ty.is_signed() => Int(*f as i64),
         (Float(f), ty) if ty.is_integral() => UnsignedInt(*f as u64),
+
+        (&Char(c), Type::Bool) => Int(c.into()),
+        (&Char(c), Type::Double) | (&Char(c), Type::Float) => Float(c.into()),
+        (&Char(c), ty) if ty.is_integral() && ty.is_signed() => Int(c.into()),
+        (&Char(c), ty) if ty.is_integral() => UnsignedInt(c.into()),
```

Comment:
What is the reason that these are references and the others are not?

Reply:
Otherwise `c` would be a `&u8` and I'd have to use `(*c).into()` which looks really ugly to my eye. I don't call `.into()` on the other variants so it's not a problem there.